### PR TITLE
Add elevator subsystem and ElevatorMove command

### PR DESCRIPTION
**Pull request by mvoodarla**
Still need to start integrating into drive code to be able to move to distinct positions.

### CODE REVIEW

File: src/main/java/org/usfirst/frc/team2473/robot/RobotMap.java [hunk 1]
```diff
@@ -11,10 +11,13 @@
 	public static final int TALON_BL = 3;
 	public static final int TALON_BR = 6;
 	public static final int TALON_FL = 2;
-	public static final int TALON_FR = 7;
+    public static final int TALON_FR = 7;
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/RobotMap.java]: Check indentation.

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+package org.usfirst.frc.team2473.robot.commands;
+
+import org.usfirst.frc.team2473.robot.Robot;
+import org.usfirst.frc.team2473.robot.RobotMap;
+import org.usfirst.frc.team2473.robot.subsystems.Elevator.ElevatorPosition;
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ * A class that enables accurate elevator mvoement to a given distance
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java]: Spelling

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		prevTicks = Robot.elevator.getEncoderTicks();
+		Robot.elevator.set(power);
+	}
+
+	@Override
+	protected void execute() {
+		double tempPower = power;
+		int currTicks = Robot.elevator.getEncoderTicks();
+		
+		int delta = currTicks - prevTicks;
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java:104]: Do you need delta for something else? If not the expression on line 94 can be simplified.

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		prevTicks = currTicks;
+		
+		Robot.elevator.printEncoders();
+		
+	}
+
+	@Override
+	protected boolean isFinished() {
+		int currTicks = Robot.elevator.getEncoderTicks();
+		if (this.moveByInches > 0) return (absoluteTickGoal < currTicks);
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java]: Please put the body of if statements on new lines, even if it's a one liner like this. It's very easy to make syntax errors in situations like this.

File: src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+import edu.wpi.first.wpilibj.command.Subsystem;
+
+/**
+ * This class contains all components of the robot necessary for driving.
+ */
+public class Elevator extends Subsystem {
+    
+    public enum ElevatorPosition {
+        BASE(100), FIRST(500), SECOND(1000), THIRD(1500);
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java]: What units are these values? How were they determined?

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorTicksTest.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import edu.wpi.first.wpilibj.command.Command;
+import org.usfirst.frc.team2473.robot.Robot;
+
+import java.io.IOException;
+
+import org.usfirst.frc.team2473.framework.Devices;
+import org.usfirst.frc.team2473.framework.UtilitySocket;
+
+/**
+ * A class that aligns the robot to the hatch based on the angle provided by CV
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorTicksTest.java:14]: https://youtu.be/BLikP6BDH5w?t=4s

File: src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    public void stop() {
+        talon.set(0);
+    }
+
+    public int getEncoderTicks() {
+        return talon.getSelectedSensorPosition(0);
+    }
+
+    public synchronized void resetEncoders() {
```
Comment (AG6GR) [inline src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java]: I'm still not sold on the use of synchronized to ensure the encoders are reset. If setSelectedSensorPosition's timeoutMs is zero, "no blocking or checking is performed." So this function will return immediately.

Also if this is truly multithreaded, encoderResetComplete should be volatile.

**Review by AG6GR — CHANGES_REQUESTED** [review]
Some minor comments/fixes.

File: src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java (hunk 5 above)
Comment (mvoodarla) [reply src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java]: Units are encoder ticks. These are just dummy values, they'll be determined when we actually get to run the elevator.

File: src/main/java/org/usfirst/frc/team2473/robot/Robot.java [hunk 8]
```diff
@@ -42,12 +46,12 @@ public void robotInit() {
 		// cvLight = new Relay(RobotMap.CV_LIGHT);
 		prefs = Preferences.getInstance();
 				
-		UsbCamera cubeCam = CameraServer.getInstance().startAutomaticCapture("Cube View", 0);
-		cubeCam.setBrightness(75);
-		cubeCam.setResolution(640, 480);
-		UsbCamera driveCam = CameraServer.getInstance().startAutomaticCapture("Front View", 1);
-		driveCam.setBrightness(75);
-		driveCam.setResolution(640, 480);
+		// UsbCamera cubeCam = CameraServer.getInstance().startAutomaticCapture("Cube View", 0);
```
Comment (mlucas) [inline src/main/java/org/usfirst/frc/team2473/robot/Robot.java]: This appears to be unrelated to this change

File: src/main/java/org/usfirst/frc/team2473/robot/Robot.java [hunk 9]
```diff
@@ -76,7 +80,7 @@ public void disabledPeriodic() {
 	 */
 	@Override
 	public void autonomousInit() {
-        Devices.getInstance().initializeCVSocket();
+        //Devices.getInstance().initializeCVSocket();
```
Comment (mlucas) [inline src/main/java/org/usfirst/frc/team2473/robot/Robot.java]: This appears to be unrelated to this change

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	/**
+	 * Lowest power elevator will run at towards the end of
+	 * the move.
+	 */
+	private static double SLOW_POWER = 0.1;
+	
+	/**
+	 * Number of inches to move by.
+	 */
+	private double moveByInches;
```
Comment (mlucas) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java]: In general, it's far easier to manage metric units as conversions are base-10.

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	/**
+	 * Create a ElevatorMove object with given inch goal and power.
+	 * @param inches inches to move by
+	 * @param power initial power to move at
+	 */
+	public ElevatorMove(double inches, double power) {
+		requires(Robot.elevator);
+		
+		this.moveByInches = inches;
+		this.power = (inches < 0) ? -power : power;
```
Comment (mlucas) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java]: What if power is negative? If you are expecting power to be > 0, you need to:
1) Assert
2) Take the absolute value

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    
+    public ElevatorMove(ElevatorPosition pos, double power) {
+		requires(Robot.elevator);
```
Comment (mlucas) [inline src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java:60]: Tabs off

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java (hunk 2 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java]: Edited, pushed this change.

File: src/main/java/org/usfirst/frc/team2473/robot/RobotMap.java (hunk 1 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/RobotMap.java]: Seems good.

File: src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java (hunk 5 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java]: Added comment, will push

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorTicksTest.java (hunk 6 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorTicksTest.java:14]: Nice 6 second YouTube video! Commented it out

File: src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java (hunk 7 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/subsystems/Elevator.java]: Edited need to do the same for DriveSubsystem as well.

File: src/main/java/org/usfirst/frc/team2473/robot/Robot.java (hunk 8 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/Robot.java]: Perhaps it was for testing?

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java (hunk 12 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java:60]: Hmm it looks fine to me on VSCode

File: src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java (hunk 3 above)
Comment (iamef) [reply src/main/java/org/usfirst/frc/team2473/robot/commands/ElevatorMove.java:104]: It just makes it easier to read.

**Comment by mvoodarla** [comment]
Replaced with #16
